dashboard-api/internal/types/studyconfig.go

```diff
@@ -0,0 +1,125 @@
+// require the same transform to be done. There is a case where some of the
+// configs are Lists (not currently implemented) where we have a case for
+func (sc *StudyConfig) TransformForDatabase() (res []DatabaseStudyConf, err error) {
+	studyID := sc.StudyID
+	v := reflect.ValueOf(*sc)
+	// we iterate through each field of the struct
+	for i := 0; i < v.NumField(); i++ {
+		// this is how we get the value of the config
+		field := v.Field(i)
+		// we use the JSON tag as that is what the  adopt
+		// service use to identify the config
+		fieldName := v.Type().Field(i).Tag.Get("json")
+		// if a field is not set we dont need to add it to our array
+		// as it will just have empty data
+		if field.IsZero() {
+			continue
+		}
+		// a simple check to see if we are dealing with a struct or a slice
+		switch field.Kind() {
+		// case where a config is of type struct
+		case reflect.Struct:
+			// we marshal the full config to json to be stored
+			conf, err := json.Marshal(field.Interface())
+			if err != nil {
+				return nil, err
+			}
+			res = append(res, DatabaseStudyConf{
+				StudyID:  studyID,
+				ConfType: fieldName,
+				Conf:     conf,
+			})
+		// case where config is a slice of configs
+		case reflect.Slice:
+			// we need to iterate through each config and add it
+			// independently when the config is a list of configs
+			for j := 0; j < field.Len(); j++ {
+				conf, err := json.Marshal(field.Index(j).Interface())
+				if err != nil {
+					return nil, err
+				}
+				res = append(res, DatabaseStudyConf{
+					StudyID:  studyID,
+					ConfType: fieldName,
+					Conf:     conf,
+				})
+			}
+		}
+	}
+	return res, nil
+}
+
+// GeneralConfig is used to hold all general configuration
+// for a study
+type GeneralConfig struct {
+	Name             string `json:"name"`
+	Objective        string `json:"objective"`
+	OptimizationGoal string `json:"optimization_goal"`
+	DestinationType  string `json:"destination_type"`
+	PageID           string `json:"page_id"`
+	MinBudget        int    `json:"min_budget"`
+	OptInWindow      int    `json:"opt_window"`
+	InstagramID      string `json:"instagram_id"`
+	AdAccount        string `json:"ad_account"`
+}
+
+// TargetingConfig describes the variables for stratification and the
+// desired joint distribution of respondents.
+type TargetingConfig struct {
+	TemplateCampaignName string          `json:"template_campaign_name"`
+	DistributionVars     DistributionVar `json:"distribution_vars"`
+}
+
+// TargetingDistributionConfig proportion of people do you want in
+// your final sample from each stratum
+type TargetingDistributionConfig struct {
+	//TODO verify if age should be a string
+	Age      string `json:"age"`
+	Gender   string `json:"gender"`
+	Location string `json:"location"`
+}
+
+// DatabaseStudyConf is the structure used to store data
+// in the database
+type DatabaseStudyConf struct {
+	CreatedAt time.Time
+	StudyID   string
+	ConfType  string
```

Comment:
So `ConfType` here refers to "General", "Targeting" etc?

In the frontend `config.type` is the form type rather than the kind of configuration that we're creating a study with, maybe I could do some renaming in the FE to change `config.type` to `form.type` to make that distinction? I don't know if it really matter though.

Reply:
Yes exactly! but I would say this shouldnt matter to the frontend as the backends job is to worry about this and do the mappings